src/Pos/Wallet/Web/Tracking.hs

```diff
@@ -201,117 +212,145 @@ syncWalletWithGStateUnsafe
     -> m ()
 syncWalletWithGStateUnsafe encSK = do
     tipHeader <- DB.getTipHeader @(Block ssc)
+    slottingData <- GS.getSlottingData
+
     let wAddr = encToCId encSK
-    whenJustM (WS.getWalletSyncTip wAddr) $ \wTip ->
-        if | wTip == genesisHash && headerHash tipHeader == genesisHash ->
-               logDebug $ sformat ("Wallet "%build%" at genesis state, synced") wAddr
-           | otherwise -> sync wAddr wTip tipHeader
-  where
-    sync :: CId Wal -> HeaderHash -> BlockHeader ssc -> m ()
-    sync wAddr wTip tipHeader = do
-        startFrom <-
-            if wTip == genesisHash then
-                resolveForwardLink wTip >>=
-                maybe (error "Unexpected state: wTip doesn't have forward link") pure
-            else pure wTip
-
-        DB.blkGetHeader startFrom >>= \case
-            Nothing ->
-                logWarning $
+        constTrue = \_ _ -> True
+        mappendR r mm = pure (r <> mm)
+        diff = (^. difficultyL)
+        mDiff = Just . diff
+        gbTxs = either (const []) (^. mainBlockTxPayload . to flattenTxPayload)
+        evalGenesisToil genUtxo = flip runUtxoReaderT genUtxo . evalToilTEmpty
+
+        mainBlkHeaderTs mBlkH =
+            getSlotStartPure True (mBlkH ^. headerSlotL) slottingData
+        blkHeaderTs = either (const Nothing) mainBlkHeaderTs
+
+        rollbackBlock :: [CWAddressMeta] -> Blund ssc -> CAccModifier
+        rollbackBlock allAddresses (b, u) =
+            trackingRollbackTxs encSK allAddresses $
+            zip3 (gbTxs b) (undoTx u) (repeat $ headerHash b)
+
+        applyBlock :: (WithLogger m1, MonadUtxoRead m1)
+                   => [CWAddressMeta] -> Blund ssc -> ToilT () m1 CAccModifier
+        applyBlock allAddresses (b, _) = trackingApplyTxs encSK allAddresses mDiff blkHeaderTs $
+                                         zip (gbTxs b) (repeat $ getBlockHeader b)
+
+        sync :: HeaderHash -> m ()
```

Comment:
i agree (only not in `when`, but `where`), but here is the problem of getting `slottingData` inside `where`-placed definition. Passing it as an argument everywhere would've been much uglier.